### ISSUE TITLE
openssl pkeyutl -verifyrecover no longer working for RSA

### ISSUE THREAD

**Issue by ifranzki** [openssl/openssl#25898]
After commit https://github.com/openssl/openssl/commit/1ee906143c0b0ebb6bcbeb87277833a665e79836 using the `openssl pkeyutl` with `-verifyrecover` using an RSA key fails with `Error: The non-raw input data length 128 is too long - max supported hashed size is 64`. That commit now also performs the check for `EVP_PKEY_OP_VERIFYRECOVER`. 

For an RSA sign processing, the input of course must be less than `EVP_MAX_MD_SIZE`. However, the signature output is then RSA modulus size (i.e. 128 bytes for an RSA 1024 bit key). 

For the subsequent `-verifyrecover` the signature is then specified with the `-in` option. From the man page https://docs.openssl.org/master/man1/openssl-pkeyutl/;

```
Recover the signed data (e.g. if an RSA key is used):

openssl pkeyutl -verifyrecover -in sig -inkey key.pem
```

So for an RSA `-verifyrecover` the input is thus the RSA modulus size, i.e. 128 bytes for an RSA 1024 bit key. This is of course larger than `EVP_MAX_MD_SIZE`, but this is totally valid for `-verifyrecover` with RSA.

I did try to use `-sigfile` instead of `-in` with `-verifyrecover`, but this results in `pkeyutl: Signature file specified for non verify`. Similar, using `-rawin` in combination with `-in` results in `pkeyutl: -digest and -rawin can only be used with -sign or -verify`.

So in my opinion, the input length should not be checked for `-verifyrecover`, only for regular sign and verify (as it was before that commit).

**Comment by ifranzki** [comment]
@DDvO fyi. This might also be related to https://github.com/openssl/openssl/pull/25831.

**Comment by DDvO** [comment]
> After commit [1ee9061](https://github.com/openssl/openssl/commit/1ee906143c0b0ebb6bcbeb87277833a665e79836) using the `openssl pkeyutl` with `-verifyrecover` using an RSA key fails with `Error: The non-raw input data length 128 is too long - max supported hashed size is 64`. That commit now also performs the check for `EVP_PKEY_OP_VERIFYRECOVER`.

Oh, right, this is a bug.

We have been confused by the `-in` option having a different meaning for `-verifyrecover` than for `-verify`.

Moreover, I just noticed that the documentation of the  `-verifyrecover` option is wrong, saying:
```
Verify the input data (which must be a hash) 
```
because, as you correctly pointed out, it must be the result of a signing operation,
and this may be longer than a digest, depending on the key type and size.

> I did try to use `-sigfile` instead of `-in` with `-verifyrecover`, but this results in `pkeyutl: Signature file specified for non verify`.

Yeah, it would have been better if `-verifyrecover` had used the `-sigfile` instead of `-in` for its signature input. 

> Similar, using `-rawin` in combination with `-in` results in `pkeyutl: -digest and -rawin can only be used with -sign or -verify`.

That's intentional and correct, because the signature out does not combine with hashing.

**Comment by DDvO** [comment]
Code fix is in #25903.

**Comment by DDvO** [comment]
Related doc fixes I noticed on this occasion have become part of #25831.
